File: sdks/python-browser-use/beater_browser_use/semconv.py
"""Canonical ``browser.*`` semantic-convention attribute keys.

These mirror the Rust source of truth in
``crates/beater-browser/src/semconv.rs``. Both the Rust capture layer and this
external instrumentation SDK emit these exact keys so the OTLP ingest mapping
(``beater-otlp``) can normalize any source identically. Keep them in lockstep
with the Rust crate.
"""

from __future__ import annotations


class SpanKind:
    """OpenInference ``openinference.span.kind`` values accepted by the normalizer.

    Browser steps project onto the existing agent-observability kinds: the LLM
    decision is an ``llm.call`` and the browser action is a ``tool.call``.
    """

    LLM_CALL = "llm.call"
    TOOL_CALL = "tool.call"


class Browser:
    """Canonical ``browser.*`` attribute keys (mirror of ``semconv.rs``)."""

    #: Browser engine that executed the step (``chromium``, ``firefox``, ...).
    ENGINE = "browser.engine"
    #: Action verb for the step (``goto``, ``click``, ``type``, ...).
    ACTION = "browser.action"
    #: Target selector for the action, when it targets an element.
    SELECTOR = "browser.selector"
    #: URL the page was on when the step executed.
    URL = "browser.url"
    #: Page title observed for the step.
    TITLE = "browser.title"
    #: Whether the action's selector resolved to an element in the DOM.
    SELECTOR_EXISTED = "browser.selector_existed"
    #: Whether the resolved element was the intended target (grounding success).
    MATCHED_ELEMENT = "browser.matched_element"
    #: Monotonic step sequence number within the agent run.
    STEP_SEQ = "browser.step_seq"
    #: Step status string (``ok`` or ``error``).
    STEP_STATUS = "browser.step_status"
    #: Artifact id of the stored DOM snapshot for the step.
    DOM_ARTIFACT = "browser.dom_artifact_id"
    #: Artifact id of the stored screenshot for the step.
    SCREENSHOT_ARTIFACT = "browser.screenshot_artifact_id"
    #: The agent's reasoning text for the decision (from ``model_thoughts``).
    REASONING = "browser.reasoning"


class Attr:
    """Non-``browser.*`` attribute keys this SDK also emits."""

    SPAN_KIND = "openinference.span.kind"
    #: Model name on the ``llm.call`` span.
    LLM_MODEL = "llm.model"


#: ``ok`` / ``error`` step-status string, matching the Rust ``StepStatus`` enum.
STATUS_OK = "ok"
STATUS_ERROR = "error"

File: sdks/ts-stagehand/src/semconv.ts
/**
 * Canonical `browser.*` semantic-convention attribute keys.
 *
 * These MUST match the Rust source of truth
 * (`crates/beater-browser/src/semconv.rs`) exactly so Beater's OTLP ingest
 * (`beater-otlp`) normalizes spans from this SDK identically to the native
 * capture layer and the other instrumentation SDKs.
 */
export const BrowserAttr = {
  /** Browser engine that executed the step (e.g. `chromium`, `firefox`, `webkit`). */
  ENGINE: "browser.engine",
  /** Action verb for the step (e.g. `act`, `observe`, `extract`). */
  ACTION: "browser.action",
  /** Target selector for the action, when the action targets an element. */
  SELECTOR: "browser.selector",
  /** URL the page was on when the step executed. */
  URL: "browser.url",
  /** Page title observed for the step. */
  TITLE: "browser.title",
  /** Whether the action's selector resolved to an element in the DOM. */
  SELECTOR_EXISTED: "browser.selector_existed",
  /** Whether the resolved element was the intended target (grounding success). */
  MATCHED_ELEMENT: "browser.matched_element",
  /** Monotonic step sequence number within the agent run. */
  STEP_SEQ: "browser.step_seq",
  /** Step status string (`ok` or `error`). */
  STEP_STATUS: "browser.step_status",
  /** Artifact id of the stored DOM snapshot for the step. */
  DOM_ARTIFACT: "browser.dom_artifact_id",
  /** Artifact id of the stored screenshot for the step. */
  SCREENSHOT_ARTIFACT: "browser.screenshot_artifact_id",
  /** The agent's reasoning text for the decision. */
  REASONING: "browser.reasoning",
} as const;

/** Step status values used for `browser.step_status`. */
export const StepStatus = {
  OK: "ok",
  ERROR: "error",
} as const;

/**
 * Beater span-kind attribute. `beater-otlp` reads this key (alongside
 * `openinference.span.kind` / `gen_ai.operation.name`) to classify the agent
 * span kind. We set it explicitly so wrapped calls map to `tool.call` and model
 * decisions map to `llm.call`.
 */
export const BEATER_SPAN_KIND = "beater.span.kind";

/** Agent span-kind values recognized by `beater-otlp`. */
export const SpanKind = {
  TOOL_CALL: "tool.call",
  LLM_CALL: "llm.call",
} as const;

/** The three Stagehand AI primitives this SDK instruments. */
export type BrowserActionName = "act" | "observe" | "extract";
